MeshGeoToolsLib/GeoMapper.cpp

```diff
@@ -352,6 +356,342 @@ void GeoMapper::advancedMapOnMesh(
+    GeoLib::Polyline::SegmentIterator& segment_it,
+    std::vector<GeoLib::LineSegment> const& sub_segments)
+{
+    std::size_t const j(segment_it.getSegmentNumber());
+    std::size_t new_pnts_cnt(0);
+    for (auto const& segment : sub_segments)
+    {
+        auto const begin_id(points.push_back(
+            new GeoLib::Point(segment.getBeginPoint(), points.size())));
+        if (ply.insertPoint(j + new_pnts_cnt + 1, begin_id))
+            new_pnts_cnt++;
+        auto const end_id(points.push_back(
+            new GeoLib::Point(segment.getEndPoint(), points.size())));
+        if (ply.insertPoint(j + new_pnts_cnt + 1, end_id))
+            new_pnts_cnt++;
+    }
+    segment_it += new_pnts_cnt;
+}
+
+static void mapPolylineOnSurfaceMesh(
+    GeoLib::Polyline& ply,
+    GeoLib::PointVec& orig_points,
+    MeshLib::MeshElementGrid const& mesh_element_grid)
+{
+    // for each segment ...
+    for (auto segment_it(ply.begin()); segment_it != ply.end(); ++segment_it)
+    {
+        auto candidate_elements(getCandidateElementsForLineSegmentIntersection(
+            mesh_element_grid, *segment_it));
+
+        auto mapPoint = [&candidate_elements](MathLib::Point3d& p) {
+            auto const* elem(
+                findElementContainingPoint2d(candidate_elements, p));
+            if (elem)
+            {
+                if (!snapPointToElementNode(p, *elem, 1e-3))
+                    mapPointOnSurfaceElement(*elem, p);
+            }
+            return elem;
+        };
+
+        // map segment begin and end point
+        auto const* beg_elem(mapPoint((*segment_it).getBeginPoint()));
+        auto const* end_elem(mapPoint((*segment_it).getEndPoint()));
+
+        // Since the mapping of the segment begin and end points the coordinates
+        // changed. The internal data structures of PointVec are possibly
+        // invalid and hence it is necessary to re-create them.
+        orig_points.resetInternalDataStructures();
+
+        if (beg_elem == end_elem)  // TODO: handle case: beg_elem == end_elem == nullptr
+            continue;
+
+        // map the line segment (and if necessary for the mapping partition it)
+        std::vector<GeoLib::LineSegment> sub_segments(mapLineSegment(
+            *segment_it, candidate_elements, beg_elem, end_elem));
+
+        if (sub_segments.empty())
+            continue;
+
+        // The case sub_segment.size() == 1 is already handled above.
+
+        if (sub_segments.size() > 1)
+        {
+            insertSubSegments(ply, orig_points, segment_it, sub_segments);
+        }
+    }
+}
+
+void GeoMapper::advancedMapOnMesh(MeshLib::Mesh const& mesh)
+{
+    // 1. extract surface
+    if (_surface_mesh)
+        delete _surface_mesh;
+    if (mesh.getDimension()<3) {
+        _surface_mesh = new MeshLib::Mesh(mesh);
+    } else {
+        const MathLib::Vector3 dir(0,0,-1);
+        _surface_mesh =
+            MeshLib::MeshSurfaceExtraction::getMeshSurface(mesh, dir, 91);
```

Comment:
OK, I also assumed something like that. You needn't change it, it only looks a bit odd. Can the `getMeshSurface()` method be adapted to have some tolerance (optional)?